sklearn/preprocessing/label.py

```diff
@@ -794,8 +795,16 @@ def _transform(self, y, class_mapping):
         """
         indices = array.array('i')
         indptr = array.array('i', [0])
+        empty_mapping = False if class_mapping else True
         for labels in y:
-            indices.extend(set(class_mapping[label] for label in labels))
```

Comment:
Hi @jnothman thanks for your comments. 
I agree that it is sometimes useful to give the user an error, but I completely disagree this is one of those cases, as the change I'm proposing only kicks in when the user EXPLICITLY has `fitted` the object and then `transforms` or given an explicit collection of `classes` to use in the object initialization and then. 
If you want to go to the real of  "when the data is problematic" it is a lost battle as there are so many exceptions that is impossible to catch in a simple pragmatic and intuitive API.

Is the responsibility of the user to check its data, and for example use the `classes` parameters to include all possible classes. Or if it is just expanding every class on every dataset then the `fit_transform` would allow that. Could you give me an example in practice you have used or seen where the solutions I'm proposing together with `fit_transform` don't work? I haven't seen one.

Your code doesn't work. I won't go to the specifics but `class_mapping` can be a `defaultdict` with a `default_factory`, which mean it increases on demand. I looked for other more compressed forms to write the code (I specially dislike line 798)  but unfortunately I couldn't find one which was simple, easy to understand, worked for `transform` and `fit_transform`, and will require changing only in one function. But if you can think of one please be my guest :)

Reply:
I think it is a good idea to raise warning on the missing class btw

Reply:
> Is the responsibility of the user to check his/her data, and for example use the classes parameters to include all possible classes.

Or it is the responsibility of the user to clean his/her data to avoid unknowns. But I do see what you mean, in a random sample context.

I suppose in this context I'm okay with changing the error to a warning. Sometimes you'd be surprised what behaviour users expect to remain backwards compatible, though! Please make that change, and we'll see what other core devs think.

> Your code doesn't work. I won't go to the specifics but class_mapping can be a defaultdict with a default_factory, which means it increases on demand.

You're right. I meant `not empty_mapping` not `class_mapping`: https://github.com/rragundez/scikit-learn/compare/mlb-unkown-class...jnothman:ignore-unknown?expand=1

Reply:
I agree that for backwards compatibility yours is the solution that bridges the gap, and I understand that even though we both might not agree on this being the best solution perhaps is the best solution for the community. So I'll add the change with under protest :)

I'll also add your piece of code, thanks!